File: 0143_reorder_list/solution.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow: ListNode = head
        fast: ListNode = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        second: ListNode = slow.next
        slow.next = None
        node: ListNode = None

        while second:
            temp: ListNode = second.next
            second.next = node
            node = second
            second = temp

        first: ListNode = head
        second: ListNode = node

        while second:
            temp1: ListNode = first.next
            temp2: ListNode = second.next
            first.next = second
            second.next = temp1
            first = temp1
            second = temp2

        

File: 0143_reorder_list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
typedef struct ListNode LN;

class Solution {
public:
    void reorderList(ListNode* head) {
        // find mid point of list
        LN *slow = head;
        LN *fast = head;
        while (fast->next && fast->next->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        // reverse 2nd half
        LN *prev = nullptr;
        LN *head2 = slow->next;
        LN *next;
        while (head2) {
            next = head2->next;
            head2->next = prev;
            prev = head2;
            head2 = next;
        }
        slow->next = nullptr;

        // merge lists
        LN *head1 = head;
        head2 = prev;
        while (head2 != nullptr) {
            next = head1->next;
            head1->next = head2;
            head1 = head2;
            head2 = next;
        }
    }
};